### ISSUE TITLE
Parameters of "empty" histograms

### ISSUE THREAD

**Issue by heliosdrm** [JuliaDynamics/RecurrenceAnalysis.jl#41]
This may happen more frequently in vertical structures, if there are no laminar states (no vertical lines, specially if a minimum line length is defined). But it might happen also with some signals for diagonal lines and recurrence times.

In such case the calculation of various parameters (averages, entropies, etc.) gives `NaN`. But is that the "right" solution? The condition is easy to check, so we may set them to zero in those cases if it's preferable. (Perhaps with a warning?)

**Comment by Datseris** [comment]
Shouldn't the return value be empty lists? This way you keep type stability as well.

**Comment by heliosdrm** [comment]
Why empty lists? Maybe you are thinking on the histograms (i.e. the return value of `recurrencestructures`). But I'm referring to the result of the RQA functions, which are numbers (integers or floats).

In this moment they are type stable:

```
julia> using RecurrenceAnalysis, SparseArrays

julia> m = sparse(Int[],Int[],Bool[],10,10)
10×10 SparseMatrixCSC{Bool,Int64} with 0 stored entries

julia> rmat = RecurrenceMatrix(m)
RecurrenceMatrix of size (10, 10) with 0 entries:

julia> det = dl_average(rmat)
NaN

julia> typeof(det)
Float64
```

The issue is to decide what result  makes sense. Using this example: if we have no lines, should the function tell that their average length cannot be calculated (e.g. implied by `NaN`, as happens now), or that it is zero?

**Comment by Datseris** [comment]
I see. Yes sorry I got confused.

I think that this case should have zero in it, not `NaN`. I also think a warning should be issued as well, for example:
```julia
@warn "We could not find vertical/diagonal lines with length greater than the given `lmin`. The returned result is zero."
```

**Comment by pucicu** [comment]
It should not be zero, when the histogram is empty. Zeros means that we would have a very discrete distribution, i.e., the histogram would have only one bin that is filled (the others are empty or zero). This is has a certain meaning. If the distribution would be completely zero for each value x, then this would be more like a uniform distribution (which would have highest entropy value). But because there is not any bin in the histogram filled, i.e., there is no bin filled, it is still different from the uniform distribution. Therefore, only NaN would be correct.

**Comment by heliosdrm** [comment]
With the current implementation, empty histograms give the following values:

* DET (for diagonal) / LAM (for vertical) : 0.0
* averages (Lmean, TT, MRT): 0.0
* maximum values (Lmax, Vmax, etc.): 0
* entropies (ENTR, RTE, etc.): 0.0
* NMPRT: 0

The only possible ambiguity that I can think of is in the case of entropies, but not in the sense that you are commenting: empty bins are not counted in the calculation of entropy, so this result is equivalent to the case where all lines or recurrent times have the same length. On the other hand, according to the formula of the Shannon entropy, zero is the correct result for an empty set, so that ambiguity is more philosophical than mathematical. For all the other parameters, those results are only possible if the histogram is empty (no lines greater than the minimum length).

We can coerce some values to `NaN` if necessary when the histogram is empty, but my gut feeling when I see `NaN`s is that something is ill, and moreover `NaN`s are propagated to all subsequent calculations, so I'd prefer to avoid them when zero may be a correct result.

Another story is how empty histograms are presented. In the current implementation of the package they are not long array fulls of zeros, but an array with only a single zero. I also thought on giving a zero-element array, but although it would work in the present state of the package, I fear that that would have a greater risk of inadverted broken code if something is changed in the future.

**Comment by pucicu** [comment]
With “distribution” I mean “frequency distribution” (e.g. histogram) and not “probability distribution”. Sorry that this was not clear.

I think, if there is not any line then the “frequency distribution” is empty and the “probability distribution” does not exist. And for something that is not existing, there cannot be a valid entropy value. Moreover, as I said, entropy = 0 would mean already something, namely one line length appearing with probability one.

BTW, the CRP Toolbox for MATLAB gives NaN in this case. 😉

**Comment by Datseris** [comment]
> We can coerce some values to NaN if necessary when the histogram is empty, but my gut feeling when I see NaNs is that something is ill, and moreover NaNs are propagated to all subsequent calculations, so I'd prefer to avoid them when zero may be a correct result.

Yeah this is a very important point. I would agree: If I ever saw a NaN everywhere I know something went terribly wrong.

**Comment by pucicu** [comment]
It should be only NaN for the entropies. Zeros for the other measures is right.

> Am 17.01.2019 um 12:39 schrieb Helios De Rosario <notifications@github.com>:
> 
> With the current implementation, empty histograms give the following values:
> 
> DET (for diagonal) / LAM (for vertical) : 0.0
> averages (Lmean, TT, MRT): 0.0
> maximum values (Lmax, Vmax, etc.): 0
> entropies (ENTR, RTE, etc.): 0.0
> NMPRT: 0
>

**Comment by Datseris** [comment]
> It should be only NaN for the entropies. Zeros for the other measures is right.

Alright, this seems okay for me as well. 

@heliosdrm 
> Another story is how empty histograms are presented. In the current implementation of the package they are not long array fulls of zeros, but an array with only a single zero. 

Wait, I am confused. Why isn't the array completely empty and instead has an entry? Shouldn't it be totally empty?

**Comment by pucicu** [comment]
But this a personal opinion. For me a NaN is not a problem. It just tells me to be more careful when further investigating the data. 🤓

**Comment by heliosdrm** [comment]
> Wait, I am confused. Why isn't the array completely empty and instead has an entry? Shouldn't it be totally empty?

It's just a consequence of how the histogram is calculated: it starts with that single "empty" bin, and then it is extended with further bins a line longer than the maximum existing bin is found. 

I prefer to keep it like that to prevent the usual out-of-bound problems with zero-element arrays. But I can modify the code of `recurrencestructures` (so far the only exported function that shows the histograms), and return `Int[]` in such cases.

**Comment by Datseris** [comment]
> It's just a consequence of how the histogram is calculated: it starts with that single "empty" bin, and then it is extended with further bins a line longer than the maximum existing bin is found.

Okay, this is fine, no need to change it.

So we all agree to only change the entropy return values to `NaN` ?

**Comment by pucicu** [comment]
I strongly suggest this. I had also discussed this in my group and they have the same opinion.

**Comment by heliosdrm** [comment]
And tell that in the documentation. :wink:
